### DISCUSSION TITLE
How to get SSR with appdir ?

### DISCUSSION THREAD

**Discussion by YoannBuzenet** [vercel/next.js#46144] [Help] [answered]
Hey there,

Switching to the new `appdir` structure, I'm losing the SSR I got with the `pages` folder. I guess I did something wrong when wrapping with Client Components ?

With appdir, my HTML code looks like that when I check it:
```html
<!DOCTYPE html>
<html id="__next_error__">
  <head>
    <script src="/_next/static/chunks/polyfills.js" nomodule=""></script>
  </head>
  <body>
    <script src="/_next/static/chunks/webpack.js" async=""></script>
    <script src="/_next/static/chunks/main-app.js" async=""></script>
  </body>
</html>
```

With javascript afterwards : 
`<script>(self.__next_f=self.__next_f||[]).push([0])</script><script>self.__next_f.push([1,"M1:{\"id\":\"./node_modules/next/dist/client/components/app-router.js\",\"name\":\"\",\"chunks\":[\"app-client-internals:app-client-internals\"],\"async\":false}\nM2:{\"id\":\"./app/providers.jsx\",\"name\":\"Providers\",\"chunks\":[\"app/layout:app/layout\"],\"async\":false}\nM3:{\"id\":\"./node_modules/next/dist/client/components/layout-router.js\",\"name\":\"\",\"chunks\":[\"app-client-internals:app-client-internals\"],\"async\":false}\nM4:{\"id\":\"./node_modules/next/dist/client/components/render-from-template-context.js\",\"name\":\"\",\"chunks\":[\"app-cl"])</script><script>self.__next_f.push([1,"ient-internals:app-client-internals\"],\"async\":false}\n"])</script><script>self.__next_f.push([1,"J0:[\"$\",\"@1\",null,{\"assetPrefix\":\"\",\"initialCanonicalUrl\":\"/posts/teslas-biggest-threat\",\"initialTree\":[\"\",{\"children\":[\"posts\",{\"children\":[[\"postURL\",\"teslas-biggest-threat\",\"oc\"],{\"children\":[\"\",{}]}]}]},null,null,true],\"initialHead\":[\"$\",\"title\",null,{\"children\":\"My TEST Next.js App\"}],\"children\":[[],[[\"$\",\"link\",\"0\",{\"rel\":\"stylesheet\",\"href\":\"/_next/static/css/app/layout.css?ts=1676839989010\",\"precedence\":\"high\"}]],[\"$\",\"html\",null,{\"lang\":\"en\",\"data-test\":\"test\",\"children\":[\"$\",\"body\",null,{\"children\":[[\"$\",\"p\",null,{\"children\":\"okkk\"}],[\"$\",\"@2\",null,{\"langHeaders\":\"fr-FR\",\"children\":[\"$\",\"@3\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\"],\"hasLoading\":false,\"template\":[\"$\",\"@4\",null,{}],\"notFound\":[\"$\",\"div\",null,{\"style\":{\"fontFamily\":\"-apple-system, BlinkMacSystemFont, Roboto, \\\"Segoe UI\\\", \\\"Fira Sans\\\", Avenir, \\\"Helvetica Neue\\\", \\\"Lucida Grande\\\", sans-serif\",\"height\":\"100vh\",\"textAlign\":\"center\",\"display\":\"flex\",\"flexDirection\":\"column\",\"alignItems\":\"center\",\"justifyContent\":\"center\"},\"children\":[[\"$\",\"head\",null,{\"children\":[\"$\",\"title\",null,{\"children\":\"404: This page could not be found.\"}]}],[\"$\",\"div\",null,{\"children\":[[\"$\",\"style\",null,{\"dangerouslySetInnerHTML\":{\"__html\":\"\\n            body { margin: 0; color: #000`

If you check, you'll see my components in this code.

If I add a component back in `pages` folder, I get it in the direct HTML, like you would expect.

I wanted to know if it is normal, or if I did something wrong. I wrapped my app in contexts, could it be the reason ? When adding just a `<p>test</p>` tag in the root layout of `appdir`, it still doesn't display in the html. I don't know where to look :)

Thank you,
Yoann

**Comment by icyJoseph** [comment, marked as answer]
Yeah this means that something failed, while rendering your page failed, or rather an error was thrown. 

Do you have a repository/branch to take a look at?

**Reply by icyJoseph** [reply]
Is this it?

https://github.com/YoannBuzenet/blog-front/blob/main/app/page.jsx#L24

`next/head` is not supported in `appDir`.

**Reply by YoannBuzenet** [reply]
That's the repo, yeah. Ok all right, I will try this and keep you posted, thank you.

**Reply by YoannBuzenet** [reply]
Hey @icyJoseph ,

I removed the `next/head` from my page component, and I still have the same behavior.

After digging, I think I found the source of the error. It's a library I made myself, and it's displaying an error server-side : window is not defined.

I thought passing in appDir would change the problem, but in fact it's still there.

`ReferenceError: window is not defined
    at 230 (webpack-internal:///(sc_client)/./node_modules/react-image-manager/dist/react-image-manager.js:2:228)
`

I removed the library and usage to see, and indeed the SSR is working. Have to find where it comes from... 
Here is the library, if you want to dig with me haha : https://github.com/YoannBuzenet/react-image-manager

Thanks for your answers !

**Reply by icyJoseph** [reply]
Ah yeah that would break. Nice, well done!
